File: demo6b-different-languages/solution/travel-planner-java/src/main/java/io/temporal/ai/workshop/travel/TravelPlannerService.java
// ABOUTME: Cross-language Nexus service interface — the Python orchestrator is the caller.
// Service/operation names and JSON field names MUST match the Python travel_planner_service.py.
package io.temporal.ai.workshop.travel;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.nexusrpc.Operation;
import io.nexusrpc.Service;

// Service name MUST equal the Python @nexusrpc.service class name "TravelPlannerService".
// (The unqualified interface name already is that; name= is stated for an explicit contract.)
// Defines the typed contract both sides of the Nexus boundary agree on — Python and Java share no code, only this interface.
@Service(name = "TravelPlannerService")
public interface TravelPlannerService {

    // JSON shape MUST match Pydantic AskRequest{ question: str }.
    class AskRequest {
        private final String question;

        @JsonCreator
        public AskRequest(@JsonProperty("question") String question) {
            this.question = question;
        }

        @JsonProperty("question")
        public String getQuestion() {
            return question;
        }
    }

    // JSON shape MUST match Pydantic AskResponse{ answer: str }.
    class AskResponse {
        private final String answer;

        @JsonCreator
        public AskResponse(@JsonProperty("answer") String answer) {
            this.answer = answer;
        }

        @JsonProperty("answer")
        public String getAnswer() {
            return answer;
        }
    }

    // Operation name MUST equal the Python operation name "ask_travel_planner".
    // The Java method default would be "askTravelPlanner", so override it explicitly.
    @Operation(name = "ask_travel_planner")
    AskResponse askTravelPlanner(AskRequest request);
}

File: demo6b-different-languages/solution/travel_planner_service.py
# ABOUTME: Caller-side Nexus service stub for the travel planner — the handler lives in Java.
# The Python orchestrator only needs this interface to build the nexus_operation_as_tool tool.

from __future__ import annotations

import nexusrpc
from pydantic import BaseModel


# Nexus operation I/O. The JSON field names ("question"/"answer") are the
# cross-language wire contract: the Java handler's Jackson POJOs must match
# these exactly. See travel-planner-java/.../TravelPlannerService.java.
class AskRequest(BaseModel):
    question: str


class AskResponse(BaseModel):
    answer: str


# Defines the typed contract both sides of the Nexus boundary agree on — Python and Java share no code, only this interface.
@nexusrpc.service
class TravelPlannerService:
    # The service name defaults to the class name ("TravelPlannerService") and
    # the operation name to this attribute name ("ask_travel_planner"). Both
    # strings must match the Java @Service / @Operation names byte-for-byte.
    ask_travel_planner: nexusrpc.Operation[AskRequest, AskResponse]
